### PR TITLE
Full dev project

### PR DESCRIPTION
**Pull request by RobinBurri**
# set up the environment of development

This branch has the environment docker for the project.
It will be the base environment for the project, but you can also use it as a sandbox when learning react or nest.
You can look at the schema in `docs/assets/docker_env_dev.png`.

* Makefile
  * **`doc` / `docdocker`** : generate `index.html` and `assets/*.svg`
  * **`start`**: launch the project
  * **`stop`**: stop the project 

### React

The React part was done with: 
```bash
npm create vite@latest 
```
and then dockerized with a node base image

### Nest

To use Nest you should install the Nest CLI on your local environment: 
> https://docs.nestjs.com/cli/overview
```bash
npm install -g @nestjs/cli
```
The Nest part was done with:
```bash
nest new project-name
```

### PostgreSQL

we add an ORM for Postgres:
```bash
npm install --save @nestjs/typeorm typeorm pg
```
we need to use env_variables that are used in docker-compose.yml so we add:
```bash
npm i --save @nestjs/config
```
then dockerized with a node base image

The Postgres part is just an image from DockerHub

### Volume

Postgres has its own name volume: data
react and nest have bind mounts to reflect changes we make while coding into the container in real-time.
(/node_modules and /dist should never be touched by us)

## All file edited by hands

```bash
find . \( -name "Makefile" \
      -or -name "*.adoc" \
      -or -name "Dockerfile" \
      -or -name ".dockerignore" \
      -or -name "docker-compose.yml" \
      -or -name "*.sh" \
      -or -name ".gitignore" \) \
     -not -path "*/node_modules/*"
```

## Try the environment

To start the project the first time you should take the following steps:

Launch transcendence project

```bash
make start # docker compose up
```

Stop running the project

```bash
make stop # docker  compose down
```

To copy node_modules from the containers to the empty ones in the project folder.
(The let the IDE know about all variables from dependencies)
```bash
sh copy-node_modules.sh
```

List the databases

```bash
docker exec -it our-postgresql psql -U $(whoami) -l
```

Enter in our database (`our-data`)

```bash
docker exec -it our-postgresql psql -d our-data -U $(whoami)
```

## Resources

* ORM: [Object–relational mapping](https://en.wikipedia.org/wiki/Object%E2%80%93relational_mapping)

### Title

```
set up the development project (#2)
```

### Description

```
list of files wrotten by hands:
find . \( -name "Makefile" \
      -or -name "*.adoc" \
      -or -name "Dockerfile" \
      -or -name ".dockerignore" \
      -or -name "docker-compose.yml" \
      -or -name "*.sh" \
      -or -name ".gitignore" \) \
     -not -path "*/node_modules/*"

enter into the database:
docker exec -it our-postgresql psql -d our-data -U $(whoami)
```

### CODE REVIEW

**Comment by t-h2o** [comment]
Is it possible to use `Nest CLI` with a docker image instead to install it on my local environment ?

File: copy-node_modules.sh [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+sudo docker cp ft_transcendence-frontend-1:/app/node_modules/. ./frontend/node_modules
+sudo docker cp ft_transcendence-backend-1:/app/node_modules/. ./backend/node_modules
```
Comment (eprei) [inline copy-node_modules.sh:4]: Ici, au lieu de 'ft_transcendence-frontend-1', ne devrait-on pas dire 'our-frontend' (comme le nom du conteneur déclaré dans le docker compose) ?

Idem pour le backend.

Comment (RobinBurri) [reply copy-node_modules.sh]: Yes, it's done.
aae1f63171b361aa3ce9bd41c20c19682ae16d52

**Review by eprei — APPROVED** [review]
Top, bravo pour tout ce travail !

**Review by eprei — APPROVED** [review]

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+services:
+    backend:
+        container_name: our-backend
+        image: our-backend-image
+        build:
+            context: ./backend/
+        ports:
+            - '8080:3000'
+        volumes:
+            - ./backend/:/app:delegated
```
Comment (t-h2o) [inline docker-compose.yml:11]: What is delegated ?
from `man 1 docker-run`
```
consistency: consistent(default), cached, or delegated. Currently, only effective for Docker for Mac.
```

Comment (RobinBurri) [reply docker-compose.yml]: old flag for optimization.
deleted.
1b03fa1a86fb4a38292240e38ebcc8f7d19890d0

**Review by t-h2o — APPROVED** [review]
LGTM

**Review by t-h2o — APPROVED** [review]
LGTM

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	@$(ADOC) $(DOCU) -o $(INDEX)
+
+docdocker:
+	@printf "$(YELLOW)launch the asciidoctor/docker-asciidoctor docker image..$(DEFAULT)\n"
+	@docker run --rm -v $(shell pwd):/documents/ asciidoctor/docker-asciidoctor make doc
+
+clean-container: clean-postgresql clean-front clean-back
+
+clean-database:
+	@if docker volume inspect $(VOLUME_DATA) 1>/dev/null 2>/dev/null ; then docker volume rm $(VOLUME_DATA); fi
```
[the remark this answers is not in the document]
Comment (RobinBurri) [reply Makefile:33]: done

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,48 @@
+version: '3.8'
+services:
+    backend:
+        container_name: our-backend
+        image: our-backend-image
+        build:
+            context: ./backend/
+        ports:
+            - '8080:3000'
```
Comment (shaolin-peanut) [inline docker-compose.yml:9]: @t-h2o je comprends pas grand chose aux ports en général, si tu peux venir m'expliquer les choix de ports pour chaque container pour clarifier un peu ce serait cool

**Review by shaolin-peanut — APPROVED** [review]
Great work, thanks for setting everything up so that we can play around and code right away 🚀

File: docker-compose.yml (hunk 4 above)
Comment (shaolin-peanut) [reply docker-compose.yml:9]: robin explained me, all good

**Pull request merged**